Fix non proper equations

FardaleM opened the pull request:
In the case of a non-proper equation `x = y` such that both variables have the same value `t`. Then, if we introduce the equation `x = y` and `y = x`, this creates a loop and gets rejected by the occur_check. The problem here is that the order of `x` and `y` is defined with respect to the value associated to it. This commit changes that to compare directly the variables. This change guaranty that the non-proper equation will form a DAG and prevent any cycle. As a side effect, this makes the comparison faster.

FardaleM committed:
Fix the insertion of non-proper equation.

In the case of a non-proper equation "x = y" such that both variables
have the same value "t". Then, if we introduce the equation "x = y" and
"y = x", this creates a loop and get rejected by the occur_check. The
problem here is that the order of "x" and "y" is defined with respect to
the value associated to it. This commit changes that to compare directly
the variables. This change guaranty that the non-proper equation will
form a DAG and prevent any cycle.
As a side effect, this makes the comparison faster.

FardaleM committed:
add test that tigger the problem with the non-proper equation

FardaleM committed:
correct non_arrow_var propagation + identical representative problem

The non_arrow_var was not properly transmitted to the attach function.
In the case of two variables, have the same representative which points to
a value, we were creating a self loop from the representative to itself.

Drup commented on lib/unification/Syntactic.ml:
```diff
@@ -212,29 +212,40 @@ and quasi_solved env stack x non_arrow s =
      Env.representative ~non_arrow:non_arrow_y env y) with
-  | (V x' | NAR x'), (V y' | NAR y') when Variable.equal x' y' -> process_stack env stack
-  | NAR y', (V x' | NAR x') | V x', NAR y' ->
+  | (V x' | NAR x' | E (x', _)), (V y' | NAR y' | E (y', _)) when Variable.equal x' y' -> process_stack env stack
```
Is the case with `V x'`, `NAR y'` with `Variable.equal x' y'` possible ? Should we assert it isn't ?

FardaleM committed:
Crash if the same variable can return as V and NAR

Env.representative should be consistent with respect to the marking of
the variable.
